File: src/targets/lib/src/multi_match.cpp
#include "REmatch/multi_match.hpp"

#include "match/match/multi_match_standard.hpp"
#include "match/match/s_multi_match.hpp"

namespace REmatch {

MultiMatch::MultiMatch(std::unique_ptr<internal::MultiMatch> match)
    : self(std::move(match)) {}

MultiMatch::MultiMatch(const MultiMatch& other)
    : self(other.self->clone()) {}

MultiMatch& MultiMatch::operator=(const MultiMatch& other) {
  if (this != &other) {
    self = other.self->clone();
  }
  return *this;
}

MultiMatch::MultiMatch(MultiMatch&& other) noexcept
    : self(std::move(other.self)) {}

MultiMatch& MultiMatch::operator=(MultiMatch&& other) noexcept {
  if (this != &other) {
    self = std::move(other.self);
  }
  return *this;
}

MultiMatch::~MultiMatch() = default;

std::vector<Span> MultiMatch::spans(const std::string& variable_name) const {
  return self->spans(variable_name);
}
std::vector<Span> MultiMatch::spans(uint_fast32_t variable_id) const {
  return self->spans(variable_id);
}
std::vector<std::string> MultiMatch::groups(const std::string& variable_name) const {
  return self->groups(variable_name);
}
std::vector<std::string> MultiMatch::groups(uint_fast32_t variable_id) const {
  return self->groups(variable_id);
}

MultiMatch MultiMatch::submatch(Span span) const {
  return MultiMatch(self->submatch(span));
}

std::vector<std::string> MultiMatch::variables() const {
  return self->variables();
}
std::string MultiMatch::to_string() const {
  return self->to_string();
}
bool MultiMatch::empty() const {
  return self->empty();
}

}  // namespace REmatch

File: src/targets/java/rematch/MultiMatch.java
package rematch;


import java.util.ArrayList;
import java.util.List;

public class MultiMatch {

    private final javacpp.MultiMatch cppMultiMatch;

    public MultiMatch(javacpp.MultiMatch cppMultiMatch) {
        this.cppMultiMatch = cppMultiMatch;
    }

    public boolean empty() {
        return cppMultiMatch.empty();
    }

    public String toString() {
        return cppMultiMatch.to_string();
    }

    public String[] variables() {
        javacpp.StringVector cppVariables = cppMultiMatch.variables();
        String[] variables = new String[(int) cppVariables.size()];
        for (int i = 0; i < cppVariables.size(); i++) {
            variables[i] = cppVariables.at(i).toString();
        }
        return variables;
    }

    public List<javacpp.Span> spans(String variableName) {
        javacpp.SpanVector spanVec = cppMultiMatch.spans(variableName);
        List<javacpp.Span> spans = new ArrayList<>();
        for (int i = 0; i < spanVec.size(); i++) {
            spans.add(spanVec.at(i));
        }
        return spans;
    }

    public List<javacpp.Span> spans(int variableId) {
        javacpp.SpanVector spanVec = cppMultiMatch.spans(variableId);
        List<javacpp.Span> spans = new ArrayList<>();
        for (int i = 0; i < spanVec.size(); i++) {
            spans.add(spanVec.at(i));
        }
        return spans;
    }

    public List<String> groups(String variableName) {
        javacpp.StringVector groupVec = cppMultiMatch.groups(variableName);
        List<String> groups = new ArrayList<>();
        for (int i = 0; i < groupVec.size(); i++) {
            groups.add(groupVec.at(i).toString());
        }
        return groups;
    }

    public List<String> groups(int variableId) {
        javacpp.StringVector groupVec = cppMultiMatch.groups(variableId);
        List<String> groups = new ArrayList<>();
        for (int i = 0; i < groupVec.size(); i++) {
            groups.add(groupVec.at(i).toString());
        }
        return groups;
    }

    public MultiMatch submatch(javacpp.Span span) {
        javacpp.MultiMatch sub = cppMultiMatch.submatch(span);
        return new MultiMatch(sub);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        MultiMatch other = (MultiMatch) obj;
        return cppMultiMatch.operatorEquals(other.cppMultiMatch);
    }
    // public boolean equals(MultiMatch other) {
    //     return cppMultiMatch.operatorEquals(other.cppMultiMatch);
    // }
}
